experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,694 @@
+end
+
+@Markdown.doc """
+    maximal_extension(
+      X::Spec{BRT, BRET, RT, RET, MST}, 
+      f::AbstractAlgebra.Generic.Frac{RET}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Returns the maximal extension of the restriction of ``f`` 
+to a rational function on ``X`` on a maximal domain of 
+definition ``U ⊂ X``. 
+"""
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    f::AbstractAlgebra.Generic.Frac{RET}
+  ) where {BRT, BRET, RT, RET, MST}
+  a = numerator(f)
+  b = denominator(f)
+  W = localized_ring(OO(X))
+  I = quotient(ideal(W, b) + localized_modulus(OO(X)), ideal(W, a))
+  U = SpecOpen(X, I)
+  g = [OO(V)(f) for V in patches(U)]
+  return StructureSheafElem(U, g)
+end
+
+@Markdown.doc """
+    maximal_extension(
+      X::Spec{BRT, BRET, RT, RET, MST}, 
+      f::Vector{AbstractAlgebra.Generic.Frac{RET}}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Returns the extension of the restriction of the ``fᵢ`` as a 
+set of rational functions on ``X`` as *regular* functions to a 
+common maximal domain of definition ``U ⊂ X``.
+"""
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    f::Vector{AbstractAlgebra.Generic.Frac{RET}}
+  ) where {BRT, BRET, RT, RET, MST}
+  if length(f) == 0
+    return SpecOpen(X), Vector{StructureSheafElem{BRT, BRET, RT, RET, MST}}()
+  end
+  R = base_ring(parent(f[1]))
+  for a in f
+    R == base_ring(parent(a)) || error("fractions do not belong to the same ring")
+  end
+  R == base_ring(OO(X)) || error("fractions do not belong to the base ring of the scheme")
+  a = numerator.(f)
+  b = denominator.(f)
+  W = localized_ring(OO(X))
+  I = ideal(W, one(W))
+  for p in f
+    I = intersect(quotient(ideal(W, denominator(p)) + localized_modulus(OO(X)), ideal(W, numerator(p))), I)
+  end
+  U = SpecOpen(X, I)
+  return U, [StructureSheafElem(U, [OO(V)(a) for V in patches(U)]) for a in f]
+end
+
+@Markdown.doc """
+    SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2}
+
+Morphisms ``f : U → V`` of open sets ``U ⊂ X`` and ``V ⊂ Y`` of affine schemes.
+These are stored as morphisms ``fᵢ: Uᵢ→ Y`` on the affine patches 
+``Uᵢ`` of ``U``.
+"""
+mutable struct SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2}
+  domain::SpecOpen{BRT, BRET, RT, RET, MST1}
+  codomain::SpecOpen{BRT, BRET, RT, RET, MST2}
+  maps_on_patches::Vector{SpecMor{BRT, BRET, RT, RET, MST1, MST2}}
+
+  # fields used for caching
+  inverse::SpecOpenMor{BRT, BRET, RT, RET, MST2, MST1}
+
+  function SpecOpenMor(
+      U::SpecOpen{BRT, BRET, RT, RET, MST1},
+      V::SpecOpen{BRT, BRET, RT, RET, MST2},
+      f::Vector{SpecMor{BRT, BRET, RT, RET, MST1, MST2}};
+      check::Bool=true
+    ) where {BRT, BRET, RT, RET, MST1, MST2}
+    Y = ambient(V)
+    n = length(f)
+    n == length(patches(U)) || error("number of patches does not coincide with the number of maps")
+    if check
+      for i in 1:n
+	domain(f[i]) == patches(U)[i] || error("domain of definition of the map does not coincide with the patch")
```

Comment:
the indentation seems off here

Reply:
Faulty presentation of tabs. But I should change my editor's settings.